geomesa-spark/geomesa-spark-sql/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSparkSQL.scala

```diff
@@ -213,31 +376,326 @@ case class GeoMesaRelation(sqlContext: SQLContext,
+  def gridIdMapper(sf: SimpleFeature, envelopes: List[Envelope]): List[(Int, SimpleFeature)] = {
+    val geom = sf.getDefaultGeometry.asInstanceOf[Geometry]
+    val mappings = envelopes.indices.flatMap { index =>
+      if (envelopes(index).intersects(geom.getEnvelopeInternal)) {
+        Some(index, sf)
+      } else {
+        None
+      }
+    }
+    if (mappings.isEmpty) {
+      List((-1, sf))
+    } else {
+      mappings.toList
+    }
+  }
+
+  // Maps a geometry to the id of the envelope that contains it
+  // Used to derive partition hints
+  def gridIdMapper(geom: Geometry, envelopes: List[Envelope]): List[Int] = {
+    val mappings = envelopes.indices.flatMap { index =>
+      if (envelopes(index).intersects(geom.getEnvelopeInternal)) {
+        Some(index)
+      } else {
+        None
+      }
+    }
+    if (mappings.isEmpty) {
+      List(-1)
+    } else {
+      mappings.toList
+    }
+  }
+
+  def spatiallyPartition(envelopes: List[Envelope], rdd: RDD[SimpleFeature], numPartitions: Int): RDD[(Int, Iterable[SimpleFeature])] = {
+    val keyedRdd = rdd.flatMap { gridIdMapper( _, envelopes)}
+    val partitioned = keyedRdd.groupByKey(new IndexPartitioner(numPartitions))
+    partitioned.foreachPartition{ iter =>
+      iter.foreach{ case (key, features) =>
+        logger.info(s"partition $key has ${features.size} features")
+      }
+    }
+    partitioned
+  }
+
+  def getBound(rdd: RDD[SimpleFeature]): Envelope = {
+    rdd.aggregate[Envelope](new Envelope())(
+      (env: Envelope, sf: SimpleFeature) => {
+        env.expandToInclude(sf.getDefaultGeometry.asInstanceOf[Geometry].getCoordinate)
+        env
+      },
+      (env1: Envelope, env2: Envelope) => {
+        env1.expandToInclude(env2)
+        env1
+      }
+    )
+  }
+
+  def equalPartitioning(bound: Envelope, numPartitions: Int): List[Envelope] = {
+    // Compute bounds of each partition
+    val partitionsPerDim = Math.sqrt(numPartitions).toInt
+    val partitionWidth = bound.getWidth / partitionsPerDim
+    val partitionHeight = bound.getHeight / partitionsPerDim
+    val minX = bound.getMinX
+    val minY = bound.getMinY
+    val partitionEnvelopes: ListBuffer[Envelope] = ListBuffer()
+
+    // Build partitions
+    for (xIndex <- 0 until partitionsPerDim) {
+      val xPartitionStart = minX + (xIndex * partitionWidth)
+      val xPartitionEnd = xPartitionStart + partitionWidth
+      for (yIndex <- 0 until partitionsPerDim) {
+        val yPartitionStart = minY + (yIndex * partitionHeight)
+        val yPartitionEnd = yPartitionStart+ partitionHeight
+        partitionEnvelopes += new Envelope(xPartitionStart, xPartitionEnd, yPartitionStart, yPartitionEnd)
+      }
+    }
+    partitionEnvelopes.toList
+  }
+
+  def weightedPartitioning(rawRDD: RDD[SimpleFeature], bound: Envelope, numPartitions: Int, sampleSize: Int): List[Envelope] = {
+    implicit val CoordinateOrdering: Ordering[Coordinate] =
+      Ordering.by {
+        case (c: Coordinate) => c.x
```

Comment:
same comment about `case` - since you're using this in multiple places maybe you could put it in a package object or something